P1957R2: remove special-case for booleans in std::variant

davidben opened the pull request:
This picks up the std::variant half of P1957R2. Fixes #62332.

davidben committed:
P1957R2: remove special-case for booleans in std::variant

This picks up the std::variant half of P1957R2. Fixes #62332.

cpplearner commented on libcxx/test/std/utilities/variant/variant.variant/variant.assign/conv.pass.cpp:
```diff
@@ -35,7 +35,7 @@ int main(int, char**)
   static_assert(!std::is_assignable<std::variant<int, const bool>, decltype("meow")>::value, "");
   static_assert(!std::is_assignable<std::variant<int, const volatile bool>, decltype("meow")>::value, "");
```
This line now causes an error in C++26 because

> error: volatile-qualified parameter type 'const volatile bool' is deprecated [-Werror,-Wdeprecated-volatile]

Since `std::variant` no longer treats `const volatile bool` specially, maybe this line could be deleted?

davidben replied:
Done.

cpplearner commented on libcxx/test/std/utilities/variant/variant.variant/variant.ctor/conv.pass.cpp:
```diff
@@ -34,7 +34,7 @@ int main(int, char**)
   static_assert(!std::is_constructible<std::variant<int, const bool>, decltype("meow")>::value, "");
   static_assert(!std::is_constructible<std::variant<int, const volatile bool>, decltype("meow")>::value, "");
```
The same error as in the `is_assignable` test:

> error: volatile-qualified parameter type 'const volatile bool' is deprecated [-Werror,-Wdeprecated-volatile]

davidben committed:
Remove const volatile bool tests

These now trigger errors in C++26:

> error: volatile-qualified parameter type 'const volatile bool' is deprecated [-Werror,-Wdeprecated-volatile]

Since std::variant no longer treats const volatile bool specially, just
remove the lines.

ldionne commented on libcxx/test/std/utilities/variant/variant.variant/variant.assign/T.pass.cpp:
```diff
@@ -145,8 +145,8 @@ void test_T_assignment_sfinae() {
                   "no boolean conversion in operator=");
-    static_assert(!std::is_assignable<V, std::false_type>::value,
-                  "no converted to bool in operator=");
+    static_assert(std::is_assignable<V, std::false_type>::value,
```
Can we add a runtime test for the assignment behavior change too?

davidben replied:
Done.

ldionne reviewed (commented):
Thanks! This looks good to me, but the CI is failing in C++26 and I'd like to add one more test for assignment.

davidben committed:
More tests

davidben committed:
Remove more C++26-incompatible volatiles

davidben commented:
> but the CI is failing in C++26

Removed a couple more `volatile`s. Running the tests locally for me isn't exercising C++26, but hopefully that's the last of it? If this still bounces, I'll give up and build Clang from trunk. :-)

ldionne reviewed (approved):
LGTM assuming CI passes -- thanks!

davidben commented:
CI is still upset, but it looks like it's unrelated? Unless I've misread the output. (On phone so a little hard to see.)

philnik777 commented:
This looks indeed unrelated.

alexfh commented:
This change may require significant cleanups, see #73121.

davidben commented:
> This change may require significant cleanups in user code, see https://github.com/llvm/llvm-project/pull/73121.

From what I can tell so far, only in repositories that set `_LIBCPP_ENABLE_NARROWING_CONVERSIONS_IN_VARIANT`. I'll follow up on that PR with a summary of the situation. It's quite a mess and I think the only long-term stable situation is to get rid of `_LIBCPP_ENABLE_NARROWING_CONVERSIONS_IN_VARIANT`.
